lib/kinesis_client/stream/coordinator.ex

```diff
@@ -146,81 +160,89 @@ defmodule KinesisClient.Stream.Coordinator do
-      end
+      add_parent_shard(graph, s)
     end)
 
     graph
   end
 
+  defp add_parent_shard(graph, %{
+         "ShardId" => shard_id,
+         "ParentShardId" => parent_shard_id,
+         "AdjacentParentShardId" => adj_parent_shard_id
+       })
+       when is_binary(parent_shard_id) do
+    add_vertex(graph, parent_shard_id)
+    :digraph.add_edge(graph, parent_shard_id, shard_id, "parent-child")
+
+    if is_binary(adj_parent_shard_id) do
+      add_vertex(graph, adj_parent_shard_id)
+      :digraph.add_edge(graph, adj_parent_shard_id, shard_id)
+    end
+  end
+
+  defp add_parent_shard(_graph, _shard), do: :ok
+
   defp start_shards(shard_graph, %__MODULE__{} = state) do
     shard_r = list_relationships(shard_graph)
 
-    Enum.reduce(shard_r, state.shard_ref_map, fn {shard_id, parents}, acc ->
-      shard_lease = get_lease(shard_id, state)
-
-      case parents do
-        [] ->
-          case shard_lease do
-            %{completed: true} ->
-              acc
-
-            _ ->
-              case start_shard(shard_id, state) do
-                {:ok, pid} -> Map.put(acc, Process.monitor(pid), shard_id)
-              end
-          end
-
-        # handle shard splits
-        [single_parent] ->
-          case get_lease(single_parent, state) do
-            %{completed: true} ->
-              Logger.info("Parent shard #{single_parent} is completed so starting #{shard_id}")
-
-              case start_shard(shard_id, state) do
-                {:ok, pid} -> Map.put(acc, Process.monitor(pid), shard_id)
-              end
-
-            %{completed: false} ->
-              Logger.info("Parent shard #{single_parent} is not completed so skipping #{shard_id}")
-              acc
-          end
-
-        # handle shard merges
-        [parent1, parent2] ->
-          case {get_lease(parent1, state), get_lease(parent2, state)} do
-            {%{completed: true}, %{completed: true}} ->
-              case start_shard(shard_id, state) do
-                {:ok, pid} -> Map.put(acc, Process.monitor(pid), shard_id)
-              end
-
-            _ ->
-              acc
-          end
-      end
+    Enum.reduce(shard_r, state.shard_pid_map, fn {shard_id, parents}, acc ->
+      maybe_start_shard(parents, shard_id, state, acc)
     end)
   end
 
+  # Shards parents are not available, so we only rely on the shard lease status
+  defp maybe_start_shard([] = _parents, shard_id, state, shard_pid_map) do
+    case get_lease(shard_id, state) do
+      # Completed shards are not started
+      %{completed: true} ->
+        shard_pid_map
+
+      _ ->
+        {:ok, shard_pid} = start_shard(shard_id, state)
+        Map.put(shard_pid_map, shard_id, shard_pid)
+    end
+  end
+
+  defp maybe_start_shard(parents, shard_id, state, shard_pid_map) do
+    if parents_completed?(parents, shard_id, state) and not shard_completed?(shard_id, state) do
+      Logger.info("Parent shards are complete and child is not, so starting #{shard_id}")
```

Comment:
Could we add a `[kcl_ex]` prefix to all the new logger calls?

While that prefix isn't consistently applied in this project, we've used it for other log lines we've added to the fork, e.g. here: https://github.com/knocklabs/kcl_ex/blob/d5a7ab30f4b3ee5293d1e878ecfaa2889ccc6ef3/lib/kinesis_client/stream/coordinator.ex#L121

This will make it a lot easier to scope log searches when debugging.

Reply:
I've added this to all kcl_ex logger calls.